compiler/rustc_mir_transform/src/add_retag.rs

```diff
@@ -50,8 +51,20 @@ fn may_be_reference(ty: Ty<'_>) -> bool {
         // References
         ty::Ref(..) => true,
         ty::Adt(..) if ty.is_box() => true,
-        // Compound types are not references
-        ty::Array(..) | ty::Slice(..) | ty::Tuple(..) | ty::Adt(..) => false,
+        // Compound types: recurse
+        ty::Array(ty, _) | ty::Slice(ty) => {
+            // This does not branch so we keep the depth the same.
```

Comment:
I don't understand the reasoning behind this. I can now ICE miri by syntactically nesting 10000 arrays, why is branching out relevant?

Reply:
This is meant to guard against time complexity, not stack overflows. Branching is a lot more expensive, if every level has N>1 fields then it is exponential in the depth, whereas array/slice checking is linear in the depth.